File: python/leetcode/Permutations.py
'''
Leetcode problem No 46 Permutations
Leetcode problem No 47 Permutations II
Solution written by Xuqiang Fang on 14 May, 2018 
'''

class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(nums, res, [])
        return res

    def dfs(self, nums, res, temp):
        if len(temp) == len(nums):
            res.append(list(temp))
            #print(res)
        for num in nums:
            if not num in temp:
                temp.append(num)
                self.dfs(nums, res, temp)
                temp.pop(len(temp)-1)
    


    '''
    for the duplicate version here, the key idea is for all duplicate items, there is an internal
    order between all the duplicate items, that way we can avoid duplicate results
    '''
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums = sorted(nums)
        used = [0] * len(nums)
        res = []
        self.helper(nums, used, res, [])
        return res

    def helper(self, nums, used, res, temp):
        if len(temp) == len(nums):
            res.append(list(temp))
            return
        for i in range(len(nums)):
            if used[i]: continue
            if i > 0 and nums[i-1] == nums[i] and (not used[i-1]): continue
            used[i] = 1
            temp.append(nums[i])
            self.helper(nums, used, res, temp)
            used[i] = 0
            temp.pop(len(temp)-1)


s = Solution()
x = [1,2,5]
print(s.permute(x))
y = [1,3,3,4,4]
z = [3,3,0,3]
print(s.permuteUnique(z))

File: java/leetcode/Permutations.java
//Leetcode problem 46 Permutations
//Leetcode problem 47 Permutations II
//Solution written by Xuqiang Fang on 14 May, 2018

import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;

class Solution{
    public List<List<Integer>> permuteUnique(int[] nums){
        List<List<Integer>> res = new ArrayList<>();
        if(nums == null || nums.length == 0)
            return res;
        boolean[] used = new boolean[nums.length];
        Arrays.sort(nums);
        dfs_unique(res, used, nums, new ArrayList<>());
        return res;
    }

    /*
    the important thing here to remember is that for the duplciate items, there should be an internal
    order between them, that way we can avoid duplicates in the final result
    */
    private void dfs_unique(List<List<Integer>> res, boolean[] used, int[] nums, List<Integer> temp){
        if(temp.size() == nums.length){
            res.add(new ArrayList<>(temp));
            return;
        } 

        for(int i=0; i<nums.length; ++i){
            if(used[i])
                continue;
            if(i > 0 && nums[i-1] == nums[i] && !used[i-1])
                continue;
            
            used[i] = true;
            temp.add(nums[i]);
            dfs_unique(res, used, nums, temp);
            used[i] = false;
            temp.remove(temp.size()-1);
        }
    }

    public List<List<Integer>> permute(int[] nums){
        List<List<Integer>> res = new ArrayList<>();
        if(nums == null || nums.length == 0)
            return res;
        boolean[] used = new boolean[nums.length];
        dfs(res, nums, new ArrayList<>(), used);
        return res;
    }
    private void dfs(List<List<Integer>> res, int[] nums, List<Integer> temp, boolean[] used){
        if(temp.size() == nums.length){
            res.add(new ArrayList<>(temp));
            return;
        }
        for(int i=0; i<nums.length; ++i){
            if(used[i])
                continue;
            used[i] = true;
            temp.add(nums[i]);
            dfs(res, nums, temp, used);
            used[i] = false;
            temp.remove(temp.size()-1);
        }
    }
}

public class Permutations{
	public static void main(String[] args){
		Solution s = new Solution();
        int[] nums = {3,3,0,3};
        int[] nums2 = {1,2,3};
        //List<List<Integer>> list = s.permuteUnique(nums);
        List<List<Integer>> list = s.permute(nums2);
        for (List<Integer> l : list)
            System.out.println(l);
	}
}
